Fix auto-update to use Lens notifications, and add update confirmation

Nokel81 opened the pull request:
- Add settings in user preferences for auto-updating (default false) and
  for allowing pre-release versions (default false)

- Use in-Lens notifications instead of OS notifications as those were
  found to be flaky

- Add rudimentary main->renderer notification system.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

a partial fix for #1768

Nokel81 committed:
Add auto-update and pre-release update user settings

- Add settings in user preferences for auto-updating (default false) and
  for allowing pre-release versions (default false)

- Use in-Lens notifications instead of OS notifications as those were
  found to be flaky

- Add rudimentary main->renderer notification system.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on src/renderer/bootstrap.tsx:
```diff
@@ -30,6 +31,10 @@ export async function bootstrap(App: AppComponent) {
   // Register additional store listeners
   clusterStore.registerIpcListener();
 
+  if (process.isMainFrame) {
```
Should this be in `LensApp`?

Nokel81 replied:
Good idea.

jakolehm commented on src/renderer/components/+preferences/preferences.tsx:
```diff
@@ -194,6 +194,24 @@ export class Preferences extends React.Component {
+          <small className="hint">
+            <Trans>Allow Lens to auto update itself to the latest version. Lens checks on startup and then once a day.</Trans>
+          </small>
+          <Checkbox
```
Pre-release should be a separate PR. I'm not sure if we really want it.

Nokel81 replied:
I'll remove it then.

jakolehm commented on src/renderer/theme.store.ts:
```diff
@@ -29,12 +29,8 @@ export class ThemeStore {
-
   @computed get activeTheme(): Theme {
-    const activeTheme = this.themes.find(theme => theme.id === this.activeThemeId) || this.themes[0];
+    const activeTheme = this.themes.find(theme => theme.id === userStore.preferences.colorTheme) || this.themes[0];
```
How are these related to this PR?

Nokel81 replied:
I did this because styling is broken if the `colorTheme` in preferences isn't valid. (And we changed the names in v4)

jakolehm commented on src/common/user-store.ts:
```diff
@@ -59,6 +61,8 @@ export class UserStore extends BaseStore<UserStoreModel> {
     colorTheme: UserStore.defaultTheme,
     downloadMirror: "default",
     downloadKubectlBinaries: true,  // Download kubectl binaries matching cluster version
+    allowAutoUpdates: false,
```
Default should be `true`.

Nokel81 replied:
Doesn't that defeat the purpose of this PR? Especially since the logic of this PR to auto update and quit immediately upon finding an update.

jakolehm replied:
Wait, I thought we should just show a notification and allow user to quit+update from there? Why would we want to quit immediately if update is found?

Nokel81 replied:
So if you look at the code, when this option is false then a notification with "Yes install now", "Yes install later", and "No" is displayed to the user. If a user wants to always install immediately then they would toggle this.

jakolehm replied:
Ok, got it... name confused me. Could we then remove this option completely (I think notification is enough)?

Nokel81 replied:
Yeah sure, I thought that we wanted auto updates (meaning installing too) which is why I added this.

Nokel81 committed:
Respond to review comments

- Remove options, always confirm, never auto prelease

- Changed "yes later" to "yes on quit"

- move register IpcHandlers

Signed-off-by: Sebastian Malton <sebastian@malton.name>

ixrock commented on src/main/app-updater.ts:
```diff
@@ -1,19 +1,170 @@
+            backchannel: noChannel,
+            accent: true
+          }
+        ],
```
Styles and translatable texts better to keep in renderer components.

Nokel81 replied:
Styles yes, I can move those. However, I don't see why we should have the text in the renderer since its source is `main`.

ixrock replied:
It doesn't matter where the "source" of texts. It's a part of layout now (especially when you show notifications in HTML-block).
For `i18n` support it's easier to keep it in `.tsx` for extracting to dictionary.
But, nevermind, localization support will required quite a lot of changes anyway.

ixrock commented on package.json:
```diff
@@ -171,13 +172,15 @@
     "chalk": "^4.1.0",
     "conf": "^7.0.1",
     "crypto-js": "^4.0.0",
+    "dateformat": "^4.3.1",
```
We have `moment` already for working with `Date` or what's the purpose to add this?

ixrock commented on src/common/user-store.ts:
```diff
@@ -2,7 +2,7 @@ import type { ThemeId } from "../renderer/theme.store";
 import semver from "semver"
 import { readFile } from "fs-extra"
-import { action, observable, reaction, toJS } from "mobx";
+import { action, IReactionDisposer, observable, reaction, toJS } from "mobx";
```
`IReactionDisposer` is non used?

Nokel81 replied:
Thanks

ixrock commented on src/main/app-updater.ts:
```diff
@@ -1,19 +1,170 @@
+    })
+
+  async function helper() {
+    while (true) {
```
`while(true){}` always looks scary :D Why this is needed?

Nokel81 replied:
I don't want to use `setTimeout` as that doesn't allow for catching errors (easily, we could use promises again I guess). This is essentially the same as calling `setTimeout` multiple times but it is more explicit. The `delay` function makes this "long polling".

jim-docker replied:
`while(true){}` is only scary when you expect execution to eventually break out of the loop

Nokel81 committed:
Responding to review comments

- use moment instead of dateformat

- remove unused code

- moved formatting notification buttons to renderer

Signed-off-by: Sebastian Malton <sebastian@malton.name>

ixrock commented on src/main/index.ts:
```diff
@@ -75,6 +73,12 @@ async function main() {
+   * This depends on:
+   * 1. windowManager: it will send IPC to the main window for notifications
+   */
+  startUpdateChecking();
```
Maybe better to `startUpdateChecking(windowManager)` for explicit dependency and no need in extra comments?

Nokel81 replied:
Agreed, though I initially had it as a comment as it was more of a transitive dependency (my function used `broadcastIpc`).

ixrock commented on src/renderer/components/notifications/notifications.store.tsx:
```diff
@@ -19,6 +23,39 @@ export interface INotification {
+      <div className="flex row align-right box grow">
+        {buttons.map(({ backchannel, ...props}) => (
+          <Button {...props} onClick={() => {
+            console.log(`sending to ${backchannel}`)
```
Forgotten `console.log`?

ixrock commented on src/renderer/components/notifications/notifications.store.tsx:
```diff
@@ -19,6 +23,39 @@ export interface INotification {
+  closeChannel?: IpcChannel;
+}
+
+function renderButtons(id: IpcChannel, buttons?: MainNotification["buttons"]): React.ReactNode {
```
Probably this should be react-component, like: `function RenderButtons(props: {id: IpcChannel, buttons?: MainNotification["buttons"]})`

ixrock commented on src/renderer/components/notifications/notifications.store.tsx:
```diff
@@ -27,6 +64,29 @@ export class NotificationsStore {
+  registerIpcListener(): void {
+    logger.info(`[NOTIFICATION-STORE] start to listen for notifications requests from main`);
+    ipcRenderer.on(NotificationChannelAdd, (event, model: MainNotification) => {
+      console.log(model);
```
console.log

ixrock reviewed (approved):
Otherwise LGTM :)

Nokel81 committed:
Resolving comments

- removed console.log

- move to RenderButtons as function component

- explicitly only send notifications to main view

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm commented on package.json:
```diff
@@ -196,6 +197,7 @@
     "mobx-observable-history": "^1.0.3",
     "mock-fs": "^4.12.0",
     "node-machine-id": "^1.1.12",
+    "node-notifier": "^9.0.0",
```
Where is this used?

Nokel81 replied:
Sorry, thought I removed all the packages I added while iterating. Will remove

Nokel81 committed:
remove unused deps

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
moment should be a normal dep

Signed-off-by: Sebastian Malton <sebastian@malton.name>

ixrock commented on package.json:
```diff
@@ -195,6 +195,7 @@
     "mobx": "^5.15.5",
     "mobx-observable-history": "^1.0.3",
     "mock-fs": "^4.12.0",
+    "moment": "^2.26.0",
```
Now, since it's required in both envs pls remove it from `devDependencies`

ixrock commented on src/main/app-updater.ts:
```diff
@@ -1,19 +1,144 @@
-export default class AppUpdater {
+function delay(duration: number): Promise<void> {
+  return new Promise(resolve => setTimeout(resolve, duration));
+}
```
Probably should be moved to `common/utils/delay.ts`

ixrock commented on src/main/app-updater.ts:
```diff
@@ -1,19 +1,144 @@
   }
+
+  helper()
+    .catch(error => logger.error("[UPDATE CHECKER]: failed with an error", { error: String(error) }));
```
So it seems we're interested only in first failure? And in case of error it will stop interval checking totally?

Nokel81 replied:
Yes that is the current behaviour, do you think we should try again if it throws an error?

jakolehm replied:
I think we should retry on next interval. Is there any reason not to retry?

Nokel81 replied:
Nope, (and I thought this too haha so it currently has that change)

Nokel81 committed:
move delay to utils, always retry even if check failed

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/main/app-updater.ts:
```diff
@@ -1,19 +1,141 @@
+    .on("update-available", async (args: UpdateInfo) => {
+      try {
+        const releaseDate = moment(args.releaseDate);
+        const body = `Version ${args.version} was release on ${releaseDate.format("dddd, mmmm dS, yyyy")}.`;
```
...was released on...

jim-docker commented on src/main/window-manager.ts:
```diff
@@ -7,7 +7,7 @@ import { initMenu } from "./menu";
 
 export class WindowManager {
-  protected mainView: BrowserWindow;
+  public readonly mainView: BrowserWindow;
```
Is WindowManager exported to extensions api? Just wondering if it is better in terms of api hardening to add a getter for `mainView` instead.

Nokel81 replied:
The generated JS (even for a getter) is no different. As is easily seen by what we do in some of our tests by casting to any (or just by using JS instead of TS).

If we really wanted to be hardened we would use the new JS `#name` private identifiers. But I don't know if we want to do that.

jim-docker replied:
So it won't break a TS extension if one day we needed to change `mainView` field back to being `protected`?

Nokel81 replied:
I don't understand, if we have `protected _mainView` and a `public get mainView()`. Then that is completely identical to `public readonly`. Modulo what the field names are.

Nokel81 committed:
fix notification rendering and disabled the auto-updater for integration tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jim-docker commented on src/main/app-updater.ts:
```diff
@@ -86,24 +87,24 @@ async function autoUpdateCheck(windowManager: WindowManager): Promise<void> {
       try {
         const releaseDate = moment(args.releaseDate);
-        const body = `Version ${args.version} was release on ${releaseDate.format("dddd, mmmm dS, yyyy")}.`;
+        const body = `Version ${args.version} was release on ${releaseDate.format("dddd, MMMM Do, yyyy")}.`;
```
```suggestion
        const body = `Version ${args.version} was released on ${releaseDate.format("dddd, MMMM Do, yyyy")}.`;
```

Nokel81 committed:
fix grammer and formatting

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
update integration runner to output logs on failure

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fixed script

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix not also failing if tests failed

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
pin minikube version

Signed-off-by: Sebastian Malton <sebastian@malton.name>
